File: Remove Duplicates from Sorted List II.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null)
            return null;
        int val = head.val - 1;
        ListNode node = head;
        head = null;        
        ListNode prev = null;
        while (node != null) {
            if (node.val == val || (node.next != null && node.val == node.next.val)) {
                // this node should be removed
                val = node.val;
                if (prev != null)
                    prev.next = node.next;
            }
            else {
                // this node is unique
                prev = node;
                if (head == null)
                    head = node;
            }
            node = node.next;
        }
        return head;
    }
}

File: Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode:
#	 def __init__(self, x):
#		 self.val = x
#		 self.next = None

class Solution:
	# @param head, a ListNode
	# @return a ListNode
	def deleteDuplicates(self, head):
		if (not head):
			return None
		node = head
		prev = None
		result = None
		preval = head.val - 1
		while (node):
			if (node.val == preval or node.next and node.next.val == node.val):
				# this node should be removed
				preval = node.val
				if (prev):
					prev.next = node.next
			else:
				# this node should be preserved (due to uniqueness)
				prev = node
				if (not result):
					result = node
			node = node.next
		return result
		